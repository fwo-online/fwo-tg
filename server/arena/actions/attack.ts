import type { Affect } from '../Constuructors/interfaces/Affect';
import PhysConstructor from '../Constuructors/PhysConstructor';
import CastError from '../errors/CastError';
/**
 * Физическая атака
 */
class Attack extends PhysConstructor implements Affect {
  constructor() {
    super({
      name: 'attack',
      displayName: 'Атака',
      desc: 'Атака',
      lvl: 0,
      orderType: 'all',
    });
  }

  /**
 * Кастомный обработчик атаки
 */
  run() {
    if (!this.params) {
      return;
    }
    const { initiator, target } = this.params;

    target.flags.isHited = {
      initiator: initiator.nick, val: this.status.effect,
    };

    target.stats.down('hp', this.status.effect);
  }

  preAffect: Affect['preAffect'] = ({ params: { initiator, target, game } }): undefined => {
    if (target.flags.isHited) {
      throw new CastError(this.getSuccessResult({ initiator: target, target: initiator, game }));
    }
  };
}
export default new Attack();
