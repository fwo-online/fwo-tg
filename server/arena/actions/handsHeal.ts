import { Heal } from '../Constuructors/HealMagicConstructor';
import type Game from '../GameService';
import type { Player } from '../PlayersService';
/**
 * Хил руками
 */
const params = {
  name: 'handsHeal',
  desc: 'Лечение руками. Действие может быть прервано физической атакой',
  displayName: 'Лечение руками',
  lvl: 0,
  orderType: 'all',
} as const;

class HandsHeal extends Heal {
  constructor() {
    super(params);
  }

  run(initiator: Player, target: Player, _game: Game): void {
    this.status.effect = this.effectVal();
    target.stats.up('hp', this.status.effect);
    target.flags.isHealed.push({
      initiator: initiator.id, val: this.status.effect,
    });
  }
}

export default new HandsHeal();
