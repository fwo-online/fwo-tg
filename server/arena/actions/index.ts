export { default as attack } from './attack';
export { default as handsHeal } from './handsHeal';
export { default as protect } from './protect';
export { default as regeneration } from './regeneration';
