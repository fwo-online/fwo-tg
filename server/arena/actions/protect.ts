import type { OrderType, ActionType } from '@/arena/Constuructors/types';
import type Game from '@/arena/GameService';
import type { Player } from '@/arena/PlayersService';
import { ProtectConstructor } from '../Constuructors/ProtectConstructor';

/**
 * Класс защиты
 */
class Protect extends ProtectConstructor {
  name = 'protect';
  displayName = 'Защита';
  desc = 'Защита от физических атак';
  lvl = 0;
  orderType: OrderType = 'all';
  actionType: ActionType = 'protect';

  run(initiator: Player, target: Player, _game: Game) {
    const protectValue = initiator.stats.val('phys.defence') * initiator.proc;
    target.stats.up('phys.defence', protectValue);
    target.flags.isProtected.push({
      initiator: initiator.id, val: protectValue,
    });
  }

  getTargetProtectors({ target } = this.params) {
    return target.flags.isProtected;
  }
}

export default new Protect();
