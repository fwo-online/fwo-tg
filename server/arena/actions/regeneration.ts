import { AffectableAction } from '../Constuructors/AffectableAction';
import type { ActionType, OrderType } from '../Constuructors/types';
import type Game from '../GameService';
import type { Player } from '../PlayersService';

/**
 * Класс регенерации
 */

class Regeneration extends AffectableAction {
  name = 'regeneration';
  displayName = 'Восстановление';
  desc = 'Регенерация маны/энергии';
  lvl = 0;
  orderType: OrderType = 'self';
  actionType: ActionType = 'regeneration';

  /**
   * Каст регенерации
   * Регенерация срабатывает даже при атаке на цель
   * @param initiator Объект кастера
   * @param target Объект цели
   * @param game Объект игры (не обязателен)
   */
  // eslint-disable-next-line class-methods-use-this
  cast(initiator: Player, target: Player, game: Game) {
    this.createContext(initiator, target, game)
    this.run(initiator, target, game)
  }

  run(initiator: Player, _target: Player, _game: Game) {
    if (initiator.prof === 'l' || initiator.prof === 'w') {
      const val = initiator.stats.val('regen.en') * initiator.proc; // размер восстан
      initiator.stats.up('en', val);
    } else {
      const val = initiator.stats.val('regen.mp') * initiator.proc; // размер восстан
      initiator.stats.up('mp', val);
    }
  }
}
export default new Regeneration();
