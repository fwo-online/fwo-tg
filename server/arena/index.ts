import type { Clan } from '@/models/clan';
import type { ItemOutput } from '@fwo/shared';
import type { ItemSet } from '@fwo/shared';
import type { CharacterService } from './CharacterService';
import type GameService from './GameService';
import type MatchMaking from './MatchMakingService';
import type * as actions from './actions';
import type * as magics from './magics';
import type * as passiveSkills from './passiveSkills';
import type * as skills from './skills';
import type * as weaponMastery from './weaponMastery';

export default {
  mm: null as unknown as typeof MatchMaking,
  characters: {} as Record<string, CharacterService>,
  games: {} as Partial<Record<string, GameService>>,
  items: {} as unknown as Record<string, ItemOutput>,
  itemsSets: {} as unknown as Record<string, ItemSet>,
  magics: {} as unknown as typeof magics,
  skills: {} as unknown as typeof skills,
  actions: {} as unknown as typeof actions &
    typeof magics &
    typeof skills &
    typeof passiveSkills &
    typeof weaponMastery,
  clans: new Map<string, Clan>(),
};
