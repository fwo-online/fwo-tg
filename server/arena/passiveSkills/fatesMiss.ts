import type { Affect } from '@/arena/Constuructors/interfaces/Affect';
import { PassiveSkillConstructor } from '@/arena/Constuructors/PassiveSkillConstructor';
import CastError from '@/arena/errors/CastError';

class FatesMiss extends PassiveSkillConstructor implements Affect {
  constructor() {
    super({
      name: 'fatesMiss',
      displayName: 'Промах судьбы',
      description: 'Даже точный удар может отклониться, если того пожелает судьба',
      chance: [5],
      effect: [0],
      bonusCost: [],
    });
  }

  run() {
    //
  }

  isActive(): boolean {
    return true;
  }

  getChance() {
    return this.chance[0];
  }

  preAffect: Affect['preAffect'] = (context): undefined => {
    this.applyContext(context);

    const { initiator, target, game } = context.params;
    if (this.checkChance()) {
      throw new CastError({
        ...this.getSuccessResult({ initiator, target, game }),
        actionType: 'miss',
      });
    }
  };
}

export default new FatesMiss();
