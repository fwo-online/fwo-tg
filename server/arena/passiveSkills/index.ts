export { default as staticProtect } from "./staticProtect";