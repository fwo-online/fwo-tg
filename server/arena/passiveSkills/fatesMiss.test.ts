import { afterEach, beforeEach, describe, expect, it } from 'bun:test';
import { attack } from '@/arena/actions';
import type GameService from '@/arena/GameService';
import TestUtils from '@/utils/testUtils';
import fatesMiss from './fatesMiss';

// npm t server/arena/passiveSkills/fatesMiss.test.ts

describe('fatesMiss', () => {
  let game: GameService;

  beforeEach(async () => {
    attack.registerPreAffects([fatesMiss]);
    game = await TestUtils.createGame([{ weapon: { type: 'stun' } }, {}]);

    TestUtils.mockRandom();
  });

  afterEach(() => {
    TestUtils.restoreRandom();
  });

  it('should break success cast', () => {
    game.players.players[0].proc = 1;
    game.players.players[0].stats.set('mp', 100);
    game.players.players[0].stats.set('magic.attack', 100);

    attack.cast(game.players.players[0], game.players.players[1], game);
    // fatesMiss.chance[0] = 100;
    TestUtils.mockRandom(0.04);
    attack.cast(game.players.players[0], game.players.players[1], game);

    expect(TestUtils.normalizeRoundHistory(game.getRoundResults())).toMatchSnapshot();
  });
});
