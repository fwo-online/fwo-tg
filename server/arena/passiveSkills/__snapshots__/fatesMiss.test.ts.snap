// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`fatesMiss should break success cast 1`] = `
"*–ò–≥—Ä–æ–∫ 1* –¥–∞–ª –ø–æ –±–∞—à–∫–µ *–ò–≥—Ä–æ–∫ 2* _–û—Ä—É–∂–∏–µ–º_ –∏ –Ω–∞–Ω—ë—Å *7.2* —É—Ä–æ–Ω–∞
\\[ –ò–≥—Ä–æ–∫ 2 üí• üíî-7.2/0.8 üìñ58 ]
*–ò–≥—Ä–æ–∫ 1* –ø—ã—Ç–∞–ª—Å—è –∞—Ç–∞–∫–æ–≤–∞—Ç—å *–ò–≥—Ä–æ–∫ 2*, –Ω–æ —É –Ω–µ–≥–æ –Ω–µ –ø–æ–ª—É—á–∏–ª–æ—Å—å
_–ü—Ä–æ–º–∞—Ö —Å—É–¥—å–±—ã_ *–ò–≥—Ä–æ–∫ 1*"
`;
